[REM] sale_margin_percentage: remove margin_percentage field in favor of native one I#25764

xmglord opened the pull request:
The field margin_percentage is no longer needed since native module
sale_margin have the field margin_percent that serve the same propose.
Also the field margin_threshold must be updated since it was used as a
percentage but the native field margin_percent is a ration so we must
follow the same pattern.

Commit:
[REM] sale_margin_percentage: remove margin_percentage field in favor of native one

The field margin_percentage is no longer needed since native module
sale_margin have the field margin_percent that serve the same propose.
Also the field margin_threshold must be updated since it was used as a
percentage but the native field margin_percent is a ration so we must
follow the same pattern.

xmglord commented:
@isaako34 and @rolandojduartem please review.

rolandojduartem commented on sale_margin_percentage/models/res_company.py:
```diff
@@ -4,4 +4,4 @@
     _inherit = "res.company"
 
-    margin_threshold = fields.Float(digits=(16, 2), help="Margin Threshold for product in order lines")
+    margin_threshold = fields.Float(digits=(18, 4), help="Margin Threshold for product in order lines")
```
Why this one?

xmglord replied:
I thought it was necessary because instead of being a percentage, 50% for example, now it must be used as a ration, 0.5 in the same example. But checking [the native field](https://github.com/odoo/odoo/blob/15.0/addons/sale_margin/models/sale_order.py#L12) I think is better to remove it.

rolandojduartem commented on sale_margin_percentage/models/sale_order.py:
```diff
@@ -1,61 +0,0 @@
-class SaleOrder(models.Model):
-    _inherit = "sale.order"
-
-    margin_percentage = fields.Float(
```
I think we need a migration script here, it is a stored field.

xmglord replied:
done

rolandojduartem commented:
There errors in the CI, is that expected?

xmglord commented:
@luisg123v please review.

luisg123v commented on sale_margin_percentage/migrations/15.0.1.0.1/post-migration.py:
```diff
@@ -0,0 +1,47 @@
+        """
+        ALTER TABLE
+            sale_order
+        DROP COLUMN IF EXISTS
```
This is not necessary, this will be done automatically by Odoo when the field is removed.

xmglord replied:
Magnificent, done.

luisg123v commented on sale_margin_percentage/tests/test_sale_order.py:
```diff
@@ -37,37 +37,37 @@ def test_01_margin_cost_0(self):
         sale = self.create_so(product=self.product_cost_10, quantity=4, price=12)
         self.assertEqual(sale.order_line.margin, 8.0)
-        self.assertAlmostEqual(sale.order_line.margin_percentage, 16.66, places=1)
+        self.assertAlmostEqual(sale.order_line.margin_percent, 0.1666, places=1)
```
Don't you need to update also places?

Same below.

xmglord replied:
done

luisg123v commented on sale_margin_percentage/tests/test_sale_order.py:
```diff
@@ -80,8 +80,9 @@ def test_07_margin_total(self):
-        )
+        calculated_margins = sale.order_line.mapped("margin_percent")
+        expected_margins = [1.0, -1.0, -0.25, 0.1667, 0.0, 0.0, 0.6154, 0.1667]
+        for expected, calculated in zip(expected_margins, calculated_margins):
```
Why?

We should avoid loops within tests

xmglord replied:
Because if I use `assertEqual` I got error since they are not computational equal and the `assertAlmostEqual` don't support list, the `numpy.testing` have a `assertAlmostEqual` that support list but I don't think use numpy is better than the loop, what do you think?

luisg123v replied:
I think you could round the computed value

Or even better, use `assertRecordValues` that will take care of rounding.

xmglord replied:
done, the `assertRecordValues`didn't take care of rounding so I rounded the computed value.

luisg123v commented on sale_margin_percentage/models/sale_order_line.py:
```diff
@@ -0,0 +1,30 @@
+from odoo import api, fields, models
+
+
+class SaleOrderLine(models.Model):
```
Could you split by model on a separate commit, please?

xmglord replied:
Done

luisg123v replied:
Commit message: it's -> its

xmglord replied:
done

Commit:
[IMP] sale_margin_percentage: separate one model per file

To follow our good practices each model should be in its own file with
the same name as the model.

luisg123v reviewed (approved):
LGTM :+1:
